### PR TITLE
Implementing ChainRulesCore.rrule for projections

### PR DESCRIPTION
**Pull request by matbesancon**

### CODE REVIEW

**Comment by matbesancon** [comment]
OK finite differences will be a pain with tolerances
https://github.com/matbesancon/MathOptSetDistances.jl/runs/1590376960?check_suite_focus=true#step:6:97

**Comment by matbesancon** [comment]
@oxinabox if you want to take a look, that's the source of all these questions :)

**Comment by matbesancon** [comment]
the interesting case here is that we have a form with all explicit derivative maps

File: test/chainrules.jl [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        for s in (MOI.Nonpositives(n), MOI.Nonnegatives(n))
+            y = MOD.projection_on_set(MOD.DefaultDistance(), x, s)
+            yb = ChainRulesTestUtils.rand_tangent(y)
+            ChainRulesTestUtils.rrule_test(
+                MOD.projection_on_set,
+                yb,
+                (MOD.DefaultDistance(), nothing),
+                (x, xb),
+                (s, nothing),
+                atol=10e-5,
```
Comment (oxinabox) [inline test/chainrules.jl:49]: this is the same as 
```suggestion
                atol=1e-4,
```
right?

File: test/chainrules.jl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                end
+                y = MOD.projection_on_set(MOD.DefaultDistance(), v, s)
+                yb = ChainRulesTestUtils.rand_tangent(y)
+                ChainRulesTestUtils.rrule_test(
+                    MOD.projection_on_set,
+                    yb,
+                    (MOD.DefaultDistance(), nothing),
+                    (v, vb),
+                    (s, nothing),
+                    atol=10e-5,
```
Comment (oxinabox) [inline test/chainrules.jl:72]:
```suggestion
                    atol=1e-4,
```

File: test/chainrules.jl [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            y = MOD.projection_on_set(MOD.DefaultDistance(), x, s)
+            yb = ChainRulesTestUtils.rand_tangent(y)
+            sb = ChainRulesTestUtils.rand_tangent(s)
+            ChainRulesTestUtils.rrule_test(
+                MOD.projection_on_set,
+                yb,
+                (MOD.DefaultDistance(), nothing),
+                (x, xb),
+                (s, sb),
+                atol=10e-5,
```
Comment (oxinabox) [inline test/chainrules.jl:97]:
```suggestion
                atol=1e-4,
```

File: test/chainrules.jl [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        end
+    end
+end
+
+@testset "rrules univariate" begin
+    @testset "Scalar $ST" for ST in (MOI.EqualTo, MOI.LessThan, MOI.GreaterThan)
+        for _ in 1:10
+            s = ST(10 * randn())
+            x = 10 * randn()
+            if isapprox(x, MOI.constant(s), atol=10e-5)
```
Comment (oxinabox) [inline test/chainrules.jl:84]:
```suggestion
            if isapprox(x, MOI.constant(s), atol=1e-5)
```

File: src/chainrules.jl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::Union{DefaultDistance, NormedEpigraphDistance}, v::AbstractVector{T}, s::MOI.SecondOrderCone) where {T}
+    vproj = projection_on_set(d, v, s)
+    t = v[1]
+    x = v[2:end]
+    norm_x = LinearAlgebra.norm2(x)
+    function pullback(Δv)
```
Comment (oxinabox) [inline src/chainrules.jl:74]: better stacktraces if you name it for the primal function
```suggestion
    function projection_on_set_pullback(Δv)
```

File: src/chainrules.jl [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        inv_sq = inv(norm_x^2)
+        cons_mul_last = inv_norm * inv_sq * t * x ⋅ Δx
+        for i in eachindex(x)
+            v̄[i+1] = inv_norm * Δt * x[i]
+            v̄[i+1] += inv_norm * (t + norm_x) * Δx[i]
+            v̄[i+1] -= cons_mul_last * x[i]
+        end
+        return result_tuple
+    end
+    return (vproj, pullback)
```
Comment (oxinabox) [inline src/chainrules.jl:97]:
```suggestion
    return (vproj, projection_on_set_pullback)
```

**Comment by oxinabox** [comment]
> OK finite differences will be a pain with tolerances
> https://github.com/matbesancon/MathOptSetDistances.jl/runs/1590376960?check_suite_focus=true#step:6:97

`isapprox(0.0, -0.09016088542119642; rtol = 1.0e-9, atol = 1.0e-9)`
that's so much error. So Much.

**Comment by matbesancon** [comment]
> isapprox(0.0, -0.09016088542119642; rtol = 1.0e-9, atol = 1.0e-9)
that's so much error. So Much.

yes I realized this one would be absurdly high. I suspect this occurs when we hit a non-differentiable point of the derivative.

I'll try to see if generating samples on the domain far enough from these is sufficient

**Pull request merged**